hw/mcu/nxp/kinetis/src/hal_timer.c

```diff
@@ -0,0 +1,428 @@
+ */
+#define MCGIRCLK_HZ 32768
+
+struct kinetis_hal_tmr {
+    LPTMR_Type *base;
+    uint16_t overflow; /* MSB of the current counter is handled manually */
+    TAILQ_HEAD(hal_timer_qhead, hal_timer) hal_timer_q;
+};
+
+#if MYNEWT_VAL(TIMER_0)
+struct kinetis_hal_tmr kinetis_tmr0 = {
+    .base = LPTMR0,
+};
+#endif
+#if MYNEWT_VAL(TIMER_1)
+struct kinetis_hal_tmr kinetis_tmr1 = {
+    .base = LPTMR1,
+};
+#endif
+
+static IRQn_Type kinetis_timer_irqs[] = LPTMR_IRQS;
+
+static struct kinetis_hal_tmr *kinetis_timers[KINETIS_TIMERS_MAX] = {
+#if MYNEWT_VAL(TIMER_0)
+    &kinetis_tmr0,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(TIMER_1)
+    &kinetis_tmr1,
+#else
+    NULL,
+#endif
+};
+
+static uint32_t
+kinetis_tmr_read(struct kinetis_hal_tmr *tmr)
+{
+    return ((uint32_t)(tmr->overflow << 16) + LPTMR_GetCurrentTimerCount(tmr->base));
+}
+
+static void
+kinetis_tmr_set_period(struct kinetis_hal_tmr *tmr, uint32_t tick)
+{
+    /* XXX only set the lower 16-bits */
+    LPTMR_SetTimerPeriod(tmr->base, tick);
+}
+
+static uint32_t
+kinetis_tmr_get_freq(struct kinetis_hal_tmr *tmr)
+{
+    uint32_t psr;
+    uint32_t freq;
+
+    psr = tmr->base->PSR;
+    switch (psr & LPTMR_PSR_PCS_MASK) {
+    case 0:
+        freq = MCGIRCLK_HZ;
+        /* if the clock divisor is not bypassed, apply the divider */
+        if (!(psr & LPTMR_PSR_PBYP_MASK)) {
+            freq >>= ((psr >> LPTMR_PSR_PRESCALE_SHIFT) & 0xffff) + 1;
+        }
+        break;
+    default:
+        /* Only MCGIRCLK is supported at the moment */
+        assert(0);
+    }
+
+    return freq;
+}
+
+static void
+kinetis_tmr_config_freq(lptmr_config_t *config, uint32_t freq_hz)
+{
+    uint32_t tmr_freq;
+    uint8_t prescaler;
+
+    config->prescalerClockSource = kLPTMR_PrescalerClock_0;
+
+    tmr_freq = MCGIRCLK_HZ;
+    prescaler = 0xff; /* overflow on purpose */
+    while (freq_hz < tmr_freq) {
+        tmr_freq /= 2;
+        prescaler++;
+        /* avoid a second overflow */
+        if (prescaler == 0xff) {
+            break;
+        }
+    }
+
+    if (tmr_freq != MCGIRCLK_HZ) {
+        config->bypassPrescaler = false;
+        config->value = prescaler;
+    } else {
+        config->bypassPrescaler = true;
+        config->value = 0;
+    }
+}
+
+static void
+timer_irq_handler(struct kinetis_hal_tmr *tmr)
+{
+    struct hal_timer *timer;
+    bool removed;
+
+    removed = false;
+    while ((timer = TAILQ_FIRST(&tmr->hal_timer_q)) != NULL) {
+        if ((int32_t)(kinetis_tmr_read(tmr) - timer->expiry) >= 0) {
+            TAILQ_REMOVE(&tmr->hal_timer_q, timer, link);
+            timer->link.tqe_prev = NULL;
+            timer->cb_func(timer->cb_arg);
+            removed = true;
+        } else {
+            break;
+        }
+    }
+
+    timer = TAILQ_FIRST(&tmr->hal_timer_q);
+    if (removed && timer) {
+        kinetis_tmr_set_period(tmr, timer->expiry);
+    } else {
+        tmr->overflow++;
```

Comment:
I'm not sure if overflow should not be handled differently.
If timer reset itself to run again in after LPTIMER overflowed, when first timer was checked again it would had value without overflow incremented so it would increment it without calling callback.
And in the next cycle timer would fire but it could be delayed while 16 bit period.

Reply:
I didn't exactly understand your use case, but I can see incrementing overflow could be skipped under some circumstances. I think I put it in an else due to some unrelated bug while debugging, and now I remove that else and moved the overflow incrementing to just before clearing the interrupt; re-tested everything (high and low speed) and it all seems to work correctly.